lib/services/spatial.js

```diff
@@ -22,36 +24,127 @@ class SpatialService {
   }
 
 
-  _destroyLayers(fourfours, core, cb) {
-    return async.map(fourfours, core.destroy.bind(core), (err, destroyResponses) => {
+  _createOrReplace(core, layer) {
+    return (layer, cb) => {
+      if (layer.uid === Layer.EMPTY) {
+        return core.create(layer, (err, resp) => {
+          cb(err, [resp, false]);
+        });
+      }
+      return core.replace(layer, (err, resp) => {
+        cb(err, [resp, true]);
+      });
+    };
+  }
+
+  /**
+   * Helper for turning a list of layers into a flat list of
+   * [layerUid, column] tuples
+   */
+  _flatColSpecs(layers) {
+    return _.flatten(layers.map((layer) => {
+      return layer.columns.map((col) => [layer.uid, col]);
+    }), true);
+  }
+
+  _destroyLayers(layers, core, cb) {
+    return async.map(layers, core.destroy.bind(core), (err, destroyResponses) => {
       return cb(err, destroyResponses);
     });
   }
 
   _createLayers(req, res, core, layers) {
-    async.map(layers, core.create.bind(core), (err, datasetResponses) => {
+    async.mapLimit(layers, MAX_PARALLEL, core.create.bind(core), (err, datasetResponses) => {
       //failed to get upstream from zk or create request failed
       if (err) return res.status(err.statusCode).send(err.body);
-      return this._createColumns(req, res, core, _.zip(layers, datasetResponses));
+      _.zip(layers, datasetResponses)
+        .forEach(([layer, response]) => layer.uid = response.body.id);
+      return this._createColumns(req, res, core, layers);
     });
   }
 
+  /**
+   * Replace the layers in core.
+   * This will
+   *   Create or replace the layers
+   *     create a layer if the id in the layer is __empty__
+   *     replace a layer if the id in the layer is a uid
+   *
+   *     These IDs are passed in from the client (core), which
+   *     ultimately is made up from the `blueprint`
+   *
+   *   Get the columns for each layer that resulted in a copy
+   *   of the schema to a new view (ie: a replace) and flatten that
+   *   into a list of [viewFourFour, columnId] tuples
+   *
+   *   Delete all those columns that were just returned, because we need
+   *   to re-create them as columns can be dropped or added
+   *
+   *   Resume the normal flow, as we now have a list of layers that have
+   *   been created in the datastore with an empty set of columns for each
+   *
+   */
+  _replaceLayers(req, res, core, layers) {
+    var createOrReplace = (layers, cb) => {
+      async.mapLimit(layers, MAX_PARALLEL, this._createOrReplace(core), (err, responses) => {
+        if (err) return cb(err);
+        var replacements = _.flatten(
+          _.zip(layers, responses)
+          .map(([layer, [response, isReplace]]) => {
+            var newUid = response.body.id;
+            if(isReplace) {
+              req.log.info(`Going to replace ${layer.uid} with new copy ${newUid}`);
```

Comment:
Nit: is it more correct to log `'Replaced ${layer.uid} with new copy ${newUid}'`?